rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExp.java

```diff
@@ -2505,38 +2678,166 @@ private static void processCharSetImpl(REGlobalData gData, RECharSet charSet) {
         if (!charSet.converted) {
             processCharSet(gData, charSet);
         }
 
+        // Check base character class membership
+        boolean matches = false;
+
         if (codePoint <= 0xFFFF) {
             int byteIndex = codePoint >> 3;
             if (!(charSet.length == 0
                     || codePoint >= charSet.length
-                    || (charSet.bits[byteIndex] & (1 << (codePoint & 0x7))) == 0))
-                return charSet.classContents.sense;
+                    || (charSet.bits[byteIndex] & (1 << (codePoint & 0x7))) == 0)) {
+                matches = true;
+            }
+        }
+
+        if (!matches && charSet.classContents.nonBMPCodepoints.contains(codePoint)) {
+            matches = true;
         }
 
-        if (charSet.classContents.nonBMPCodepoints.contains(codePoint))
-            return charSet.classContents.sense;
+        if (!matches) {
+            for (int i = 0; i < charSet.classContents.nonBMPRanges.size(); i += 2) {
+                if (codePoint >= charSet.classContents.nonBMPRanges.get(i)
+                        && codePoint <= charSet.classContents.nonBMPRanges.get(i + 1)) {
+                    matches = true;
+                    break;
+                }
+            }
+        }
+
+        if (!matches) {
+            for (int encodedProp : charSet.unicodeProps) {
+                if (UnicodeProperties.hasProperty(encodedProp, codePoint)) {
+                    matches = true;
+                    break;
+                }
+            }
+        }
+
+        if (!matches) {
+            for (int encodedProp : charSet.negUnicodeProps) {
+                if (!UnicodeProperties.hasProperty(encodedProp, codePoint)) {
+                    matches = true;
+                    break;
+                }
+            }
+        }
+
+        // Apply set operations for 'v' flag
+        if (!charSet.classContents.setOperations.isEmpty()) {
+            for (SetOperation op : charSet.classContents.setOperations) {
+                boolean operandMatches = checkClassContentsMatch(gData, op.operand, codePoint);
+
+                if (op.type == SetOperationType.SUBTRACT) {
+                    // Remove from result if in subtract operand
+                    if (operandMatches) {
+                        matches = false;
+                    }
+                } else if (op.type == SetOperationType.INTERSECT) {
+                    // Keep only if also in intersect operand
+                    matches = matches && operandMatches;
+                }
+            }
+        }
+
+        return matches == charSet.classContents.sense;
+    }
+
+    // Helper method to check if codePoint matches a ClassContents
+    private static boolean checkClassContentsMatch(
+            REGlobalData gData, ClassContents contents, int codePoint) {
+        // Build a temporary RECharSet for the operand
+        RECharSet tempCharSet = new RECharSet(contents, 65536); // Standard BMP length
```

Comment:
This temporary RECharSet is built each time the regexp is executed right ? Any reason this can't be done just once ?